src/main/java/Parser.java

```diff
@@ -0,0 +1,153 @@
+     * @return index position or 0
+     */
+    private int getPrepositionPos(String[] inputSplit) {
+        for (int i = 0; i < inputSplit.length; i++) {
+            if (inputSplit[i].charAt(0) == '/') {
+                return i;
+            }
+        }
+        return 0;
+    }
+
+    /**
+     * Returns the Task constructed from the input String array
+     * @param inputSplit given String array
+     * @return Task generated
+     * @throws InvalidInputFormatException If the input is not in the required format
+     */
+    private Task construct(String[] inputSplit) throws InvalidInputFormatException {
+        if (inputSplit.length < 2) {
+            throw new InvalidInputFormatException();
+        }
+        checkValidSpacing(inputSplit);
+        Task task;
+        int prepAt = getPrepositionPos(inputSplit);
+        StringBuilder description = new StringBuilder();
+        String preposition = inputSplit[prepAt].substring(1);
+        StringBuilder memo = new StringBuilder();
+        boolean prepRequired = inputSplit[0].equals("deadline") || inputSplit[0].equals("event");
+
+        if (prepRequired && prepAt == 0) {
+            // System.out.println("what's the date due?");
+            throw new InvalidInputFormatException();
+        }
+        if (prepRequired) {
+            for (int k = 1; k < prepAt; k++) {
+                description.append(" ").append(inputSplit[k]);
+            }
+        } else {
+            for (int k = 1; k < inputSplit.length; k++) {
+                description.append(" ").append(inputSplit[k]);
+            }
+        }
+        for (int i = prepAt + 1; i < inputSplit.length; i++) {
+            memo.append(" ").append(inputSplit[i]);
+        }
+
+        switch (inputSplit[0]) {
+        case "todo":
+            task = new ToDo(description.deleteCharAt(0).toString());
+            break;
+        case "deadline":
+            task = new Deadline(description.deleteCharAt(0).toString(), preposition, memo.deleteCharAt(0).toString());
+            break;
+        case "event":
+            task = new Event(description.deleteCharAt(0).toString(), preposition, memo.deleteCharAt(0).toString());
+            break;
+        default:
+            task = null;
+        }
+        return task;
+    }
+
+    void checkValidTaskIndex(String[] inputSplit) throws InvalidInputFormatException {
+        if (inputSplit.length < 2) {
+            throw new InvalidInputFormatException();
+        }
+        try {
+            Integer.parseInt(inputSplit[1]);
+        } catch (NumberFormatException e) {
+            throw new InvalidInputFormatException();
+        }
+    }
+
+    void checkValidSpacing(String[] inputSplit) throws InvalidInputFormatException {
+        for (String s : inputSplit) {
+            if (s.length() == 0) {
+                throw new InvalidInputFormatException();
+            }
+        }
+    }
+
+    boolean parse(String input, TaskList tasks, Ui ui, Storage storage) {
```

Comment:
It seems that your parser handles logic of ui and storage too. May I suggest this alternative POV - 
Parser to only handle the breaking down of the string; ui and storage can be handled in the `run()` method of Duke. This lies more in line with the defn of parsing.

Reply:
Yeah thanks I didn't think of that. Now that I saw your code, returning commands from parse makes it more intuitive. I think I'll do that for the group project.